### PR TITLE
Support full-text search

### PR DESCRIPTION
**Pull request by sunxiaoguang**
Add MATCH(col1,col2,...) AGAINST(expr [search_modifier]) syntax support

Signed-off-by: Xiaoguang Sun <sunxiaoguang@zhihu.com>

### What problem does this PR solve? 
Support full-text search syntax

### What is changed and how it works?
Add MATCH(col1,col2,...) AGAINST(expr [search_modifier]) syntax support

### Check List 

Tests 
 - Unit test
New tests added to expression_test.go and parser_test.go 

Code changes
 - Has exported function/method change
NO
 - Has exported variable/fields change
NO 
- Has interface methods change
NO

Side effects
NA

Related changes
NA

### CODE REVIEW

**Comment by sunxiaoguang** [comment]
The coverage test seems to be hard to fix. Testing the error path depends on existing code to return error.

File: ast/expressions.go [hunk 1]
```diff
@@ -1273,3 +1273,56 @@ func (n *MaxValueExpr) Accept(v Visitor) (Node, bool) {
...
+	}
+	ctx.WritePlain(") ")
+	ctx.WriteKeyWord("AGAINST")
+	ctx.WritePlain(" (")
+	if err := n.Against.Restore(ctx); err != nil {
+		return errors.Annotate(err, "An error occurred while restore MatchAgainst.Against")
+	}
+	if n.Modifier&FulltextSearchModifierNaturalLanguageMode == FulltextSearchModifierNaturalLanguageMode {
+		ctx.WritePlain(" IN NATURAL LANGUAGE MODE")
+	}
```
Comment (kennytm) [inline ast/expressions.go]: ... and don't print `IN NATURAL LANGUAGE MODE` in the restored text as it is redundant

File: ast/dml.go [hunk 2]
```diff
@@ -2605,3 +2605,11 @@ func (n *SplitOption) Restore(ctx *RestoreCtx) error {
 	}
 	return nil
 }
+
+type FulltextSearchModifier int
+
+const (
+	FulltextSearchModifierNaturalLanguageMode = 1 << iota
```
[the remark this answers is not in the document]
Comment (sunxiaoguang) [reply ast/dml.go]: Makes sense, modes do not mix.

[the remark this answers is not in the document]
Comment (sunxiaoguang) [reply ast/dml.go]: Updated, please talk a look.

File: ast/dml.go [hunk 3]
```diff
@@ -2605,3 +2605,24 @@ func (n *SplitOption) Restore(ctx *RestoreCtx) error {
...
 	return nil
 }
+
+type FulltextSearchModifier int
+
+const (
+	FulltextSearchModifierNaturalLanguageMode = 0
+	FulltextSearchModifierBooleanMode         = 1
+	FulltextSearchModifierModeMask            = 0xF
+	FulltextSearchModifierWithQueryExpansion  = 1 << 4
```
Comment (kennytm) [inline ast/dml.go]: (why `1 << 4`? 😅)

File: parser.y [hunk 4]
```diff
@@ -3898,6 +3902,14 @@ Expression:
 			$$ = &ast.UnaryOperationExpr{Op: opcode.Not, V: $2}
 		}
 	}
+|	"MATCH" '(' ColumnNameList ')' "AGAINST" '(' StringLiteral FulltextSearchModifierOpt ')'
```
Comment (kennytm) [inline parser.y]: Wait, the AGAINST part can be an arbitrary (string) expression, not just string literal.

```
mysql> select match(a) against (1234), match(a) against (now()), match(a) against (@@sql_mode in boolean mode) from x;
+-------------------------+--------------------------+-----------------------------------------------+
| match(a) against (1234) | match(a) against (now()) | match(a) against (@@sql_mode in boolean mode) |
+-------------------------+--------------------------+-----------------------------------------------+
|                       0 |                        0 |                                             0 |
|                       0 |                        0 |                                             0 |
+-------------------------+--------------------------+-----------------------------------------------+
2 rows in set (0.00 sec)
```

File: ast/dml.go (hunk 3 above)
Comment (sunxiaoguang) [reply ast/dml.go]: In case we want to support more modes compare to MySQL later on.

File: parser.y (hunk 4 above)
Comment (sunxiaoguang) [reply parser.y]: Hm, it looks strange to query from variable though.

Comment (kennytm) [reply parser.y]: I think it's better to support it for compatibility with MySQL. I guess const-evaluating the expression in TiDB isn't difficult.

Comment (sunxiaoguang) [reply parser.y]: Agree, let me fix it.

Comment (sunxiaoguang) [reply parser.y]: @kennytm

File: ast/expressions.go [hunk 5]
```diff
@@ -1273,3 +1274,70 @@ func (n *MaxValueExpr) Accept(v Visitor) (Node, bool) {
...
+	fmt.Fprint(w, ")")
+}
+
+func (n *MatchAgainst) Accept(v Visitor) (Node, bool) {
+	newNode, skipChildren := v.Enter(n)
+	if skipChildren {
+		return v.Leave(newNode)
+	}
+	return v.Leave(n)
+}
```
[the remark this answers is not in the document]
Comment (sunxiaoguang) [reply ast/expressions.go]: Sure, let me try to understand what Accept does 😂

**Comment by sunxiaoguang** [comment]
> LGTM

Thanks, But what does codecov/patch mean? I have no idea what is it about.

**Comment by kennytm** [comment]
@sunxiaoguang it means the coverage of the _new_ code is not enough by codecov's standard. I won't worry about that too much since it is not required to pass.

**Comment by kennytm** [comment]
PTAL @tiancaiamao or @tangenta

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**